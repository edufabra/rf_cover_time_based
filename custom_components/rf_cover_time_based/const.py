"""Constants for the RF Cover Time Based integration."""
from __future__ import annotations

from typing import Final

# This is the domain of your integration. It must be unique.
DOMAIN: Final = "rf_cover_time_based"

# The version of the integration.
VERSION: Final = "0.1.0"

#Configuration keys used throughout the integration
CONF_NAME = "name"
CONF_REMOTE_ENTITY = "remote_entity"
CONF_TRAVELLING_TIME_DOWN = "travelling_time_down"
CONF_TRAVELLING_TIME_UP = "travelling_time_up"
CONF_OPEN_COMMAND = "open_command"
CONF_CLOSE_COMMAND = "close_command"
CONF_STOP_COMMAND = "stop_command"
CONF_DEVICE_CLASS = "device_class"
