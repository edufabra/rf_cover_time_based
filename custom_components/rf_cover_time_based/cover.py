"""The cover platform for the RF Cover Time Based integration."""
from __future__ import annotations

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

# Import the actual entity class from your main implementation file.
from .time_based_cover import TimeBasedCover


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """
    Set up the cover entity from a config entry.

    This function is called by Home Assistant to set up the cover platform.
    It's responsible for creating and adding the cover entity.
    """
    # The fix is here: Pass both `hass` and `config_entry` to the constructor,
    # which resolves the TypeError seen in the test logs.
    async_add_entities([TimeBasedCover(hass, config_entry)])
