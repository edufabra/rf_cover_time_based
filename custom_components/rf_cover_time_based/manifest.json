{
  "domain": "rf_cover_time_based",
  "name": "RF Cover Time Based",
  "version": "2025.7.4",
  "documentation": "https://github.com/edufabra/rf_cover_time_based",
  "issue_tracker": "https://github.com/edufabra/rf_cover_time_based/issues",
  "codeowners": ["@edufabra"],
  "config_flow": true,
  "integration_type": "entity",
  "dependencies": ["remote"],
  "after_dependencies": ["broadlink", "tuya"],
  "iot_class": "calculated",
  "quality_scale": "platinum"
}
