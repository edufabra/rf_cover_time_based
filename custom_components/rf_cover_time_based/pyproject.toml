[tool.ruff]
line-length = 88
indent-width = 4

[tool.ruff.lint]
select = [
    "E",  # pycodestyle errors
    "W",  # pycodestyle warnings
    "F",  # pyflakes
    "I",  # isort
    "C",  # flake8-comprehensions
    "B",  # flake8-bugbear
]

[tool.ruff.lint.isort]
known-first-party = ["homeassistant", "custom_components"]