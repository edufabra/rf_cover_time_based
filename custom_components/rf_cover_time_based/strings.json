{
  "config": {
    "step": {
      "user": {
        "title": "RF Time Based Cover Setup",
        "description": "Enter the details for your time-based cover.",
        "data": {
          "name": "Name",
          "remote_entity": "Remote Entity",
          "open_command": "Open Command",
          "close_command": "Close Command",
          "stop_command": "Stop Command",
          "travelling_time_down": "Travel Time Down (seconds)",
          "travelling_time_up": "Travel Time Up (seconds)",
          "device_class": "Device Class"
        }
      }
    },
    "error": {
      "cannot_connect": "Failed to connect to the remote entity. Please check the entity ID.",
      "unknown": "An unknown error occurred."
    },
    "abort": {
      "already_configured": "This device is already configured."
    }
  },
  "options": {
    "step": {
      "init": {
        "title": "RF Time Based Cover Options",
        "description": "Update the settings for your time-based cover.",
        "data": {
          "remote_entity": "Remote Entity",
          "open_command": "Open Command",
          "close_command": "Close Command",
          "stop_command": "Stop Command",
          "travelling_time_down": "Travel Time Down (seconds)",
          "travelling_time_up": "Travel Time Up (seconds)",
          "device_class": "Device Class"
        }
      }
    }
  }
}