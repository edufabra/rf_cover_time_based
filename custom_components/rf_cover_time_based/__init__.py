# custom_components/rf_cover_time_based/__init__.py

"""The RF Cover Time Based integration."""
from __future__ import annotations

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant

# Define the platforms that this integration will create.
PLATFORMS: list[Platform] = [Platform.COVER]

async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up from a config entry."""
    # This is the central point to forward the setup to the platforms.
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    # Add an update listener to the config entry for options flow.
    entry.async_on_unload(entry.add_update_listener(async_reload_entry))

    return True

async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    # This is the central point to forward the unload to the platforms.
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)


async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Handle an options update by reloading the entry."""
    await hass.config_entries.async_reload(entry.entry_id)
