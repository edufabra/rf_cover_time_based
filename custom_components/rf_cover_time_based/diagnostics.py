"""Diagnostics support for RF Cover Time Based."""
from __future__ import annotations

from typing import Any

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, State
from homeassistant.helpers.device_registry import (
    async_get as async_get_device_registry,
)
from homeassistant.helpers.entity_registry import (
    EntityRegistry,
)
from homeassistant.helpers.entity_registry import (
    async_get as async_get_entity_registry,
)

from .const import CONF_REMOTE_ENTITY, DOMAIN


def _get_entity_diagnostic_data(entity_state: State | None) -> dict[str, Any]:
    """Return a standardized diagnostic dictionary for a given entity state."""
    if not entity_state:
        return {"state": "not_found", "attributes": {}}

    return {
        "state": entity_state.state,
        "attributes": dict(entity_state.attributes),
    }


def _get_redacted_config_entry(entry: ConfigEntry) -> dict[str, Any]:
    """
    Return a redacted config entry dictionary for diagnostics.

    This removes sensitive "command" keys for privacy.
    """
    entry_dict = entry.as_dict()
    if "data" in entry_dict:
        entry_dict["data"] = {
            key: value
            for key, value in entry_dict["data"].items()
            if "command" not in key
        }
    if "options" in entry_dict:
        entry_dict["options"] = {
            key: value
            for key, value in entry_dict["options"].items()
            if "command" not in key
        }
    return entry_dict


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    device_registry = async_get_device_registry(hass)
    entity_registry = async_get_entity_registry(hass)

    config = {**entry.data, **entry.options}

    device = device_registry.async_get_device(identifiers={(DOMAIN, entry.entry_id)})

    cover_entity_id = _find_entity_for_device(
        entity_registry, device.id if device else None, "cover"
    )
    cover_entity_state = (
        hass.states.get(cover_entity_id) if cover_entity_id else None
    )

    remote_entity_id = config.get(CONF_REMOTE_ENTITY)
    remote_entity_state = (
        hass.states.get(remote_entity_id) if remote_entity_id else None
    )

    return {
        "config_entry": _get_redacted_config_entry(entry),
        "cover_entity": {
            "entity_id": cover_entity_id,
            **_get_entity_diagnostic_data(cover_entity_state),
        },
        "remote_gateway": {
            "entity_id": remote_entity_id,
            **_get_entity_diagnostic_data(remote_entity_state),
        },
    }


def _find_entity_for_device(
    entity_registry: EntityRegistry, device_id: str | None, platform: str
) -> str | None:
    """Find the first entity of a given platform for a specific device."""
    if not device_id:
        return None

    for entity in entity_registry.entities.values():
        if entity.device_id == device_id and entity.platform == platform:
            return entity.entity_id

    return None
