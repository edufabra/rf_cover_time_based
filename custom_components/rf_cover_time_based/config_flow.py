# custom_components/rf_cover_time_based/config_flow.py
"""Config flow for RF Cover Time Based integration."""
from __future__ import annotations

import logging
from typing import Any

import voluptuous as vol

from homeassistant import config_entries
from homeassistant.components.cover import CoverDeviceClass
from homeassistant.core import callback
from homeassistant.data_entry_flow import FlowResult
from homeassistant.helpers.selector import (
    EntitySelector,
    EntitySelectorConfig,
    SelectSelector,
    SelectSelectorConfig,
    SelectSelectorMode,
)

from .const import (
    CONF_CLOSE_COMMAND,
    CONF_DEVICE_CLASS,
    CONF_NAME,
    CONF_OPEN_COMMAND,
    CONF_REMOTE_ENTITY,
    CONF_STOP_COMMAND,
    CONF_TRAVELLING_TIME_DOWN,
    CONF_TRAVELLING_TIME_UP,
    DOMAIN,
)

_LOGGER = logging.getLogger(__name__)


def _build_options_schema(options: dict[str, Any]) -> vol.Schema:
    """Build the schema for the options form, pre-populating with existing values."""
    return vol.Schema(
        {
            vol.Required(
                CONF_REMOTE_ENTITY,
                default=options.get(CONF_REMOTE_ENTITY),
            ): EntitySelector(EntitySelectorConfig(domain="remote")),
            vol.Required(
                CONF_OPEN_COMMAND, default=options.get(CONF_OPEN_COMMAND)
            ): str,
            vol.Required(
                CONF_CLOSE_COMMAND,
                default=options.get(CONF_CLOSE_COMMAND),
            ): str,
            vol.Required(
                CONF_STOP_COMMAND, default=options.get(CONF_STOP_COMMAND)
            ): str,
            vol.Required(
                CONF_TRAVELLING_TIME_DOWN,
                default=options.get(CONF_TRAVELLING_TIME_DOWN, 10),
            ): int,
            vol.Required(
                CONF_TRAVELLING_TIME_UP,
                default=options.get(CONF_TRAVELLING_TIME_UP, 10),
            ): int,
            vol.Required(
                CONF_DEVICE_CLASS,
                default=options.get(CONF_DEVICE_CLASS, "shutter"),
            ): SelectSelector(
                SelectSelectorConfig(
                    options=[cls.value for cls in CoverDeviceClass],
                    mode=SelectSelectorMode.DROPDOWN,
                )
            ),
        }
    )


class RfCoverTimeBasedConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a config flow for RF Cover Time Based."""

    VERSION = 1

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Handle the initial step."""
        if not self.hass.states.async_all("remote"):
            return self.async_abort(reason="no_remotes_found")

        if user_input is not None:
            # Set the unique_id based on the name to prevent duplicates.
            await self.async_set_unique_id(user_input[CONF_NAME])
            self._abort_if_unique_id_configured()

            # Separate the configuration into data (immutable) and options (mutable).
            name = user_input.pop(CONF_NAME)

            # The 'data' dictionary should be minimal.
            config_data = {}

            # The rest of the user input becomes the initial options.
            config_options = user_input

            return self.async_create_entry(
                title=name, data=config_data, options=config_options
            )

        # Build the initial user schema, which includes the name field
        user_schema = vol.Schema({vol.Required(CONF_NAME): str}).extend(
            _build_options_schema({}).schema
        )

        return self.async_show_form(step_id="user", data_schema=user_schema)

    @staticmethod
    @callback
    def async_get_options_flow(
        config_entry: config_entries.ConfigEntry,
    ) -> config_entries.OptionsFlow:
        """Get the options flow for this handler."""
        return RfCoverTimeBasedOptionsFlow(config_entry)


class RfCoverTimeBasedOptionsFlow(config_entries.OptionsFlowWithConfigEntry):
    """Handle an options flow for RF Cover Time Based."""

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Manage the options."""
        if user_input is not None:
            # This creates an entry in the `options` dictionary of the ConfigEntry
            return self.async_create_entry(title="", data=user_input)

        # Reuse the schema builder, passing the existing options
        options_schema = _build_options_schema(self.options)

        return self.async_show_form(step_id="init", data_schema=options_schema)