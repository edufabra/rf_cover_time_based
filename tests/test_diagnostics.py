"""Test the RF Cover Time Based diagnostics."""
from homeassistant.core import HomeAssistant
from pytest_homeassistant_custom_component.common import MockConfigEntry
from syrupy import SnapshotAssertion

from custom_components.rf_cover_time_based.diagnostics import (
    async_get_config_entry_diagnostics,
)


async def test_entry_diagnostics(
    hass: HomeAssistant,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test config entry diagnostics."""
    # The init_integration fixture is already setting up the entry.
    # We can directly call the diagnostics function.
    diagnostics_data = await async_get_config_entry_diagnostics(
        hass, init_integration
    )

    # Use syrupy snapshot testing to verify the output.
    # This creates a snapshot file on the first run, and compares against it
    # on subsequent runs. It's a powerful way to test complex data structures.
    assert diagnostics_data == snapshot
