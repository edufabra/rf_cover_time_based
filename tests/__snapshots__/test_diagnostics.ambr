# serializer version: 1
# name: test_entry_diagnostics
  dict({
    'config_entry': dict({
      'data': dict({
        'device_class': 'shutter',
        'name': 'Test Shutter',
        'remote_entity': 'remote.test_gateway',
        'travelling_time_down': 10,
        'travelling_time_up': 10,
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'rf_cover_time_based',
      'entry_id': 'test-shutter',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'Mock Title',
      'unique_id': None,
      'version': 1,
    }),
    'cover_entity': dict({
      'attributes': dict({
      }),
      'entity_id': None,
      'state': 'not_found',
    }),
    'remote_gateway': dict({
      'attributes': dict({
      }),
      'entity_id': 'remote.test_gateway',
      'state': 'not_found',
    }),
  })
# ---
